app/code/Magento/InventoryElasticsearch/Plugin/CatalogSearch/Model/Indexer/Fulltext/Action/DataProvider/StockedProductFilterByInventoryStock.php

```diff
@@ -0,0 +1,143 @@
+     */
+    private $stockByWebsiteIdResolver;
+
+    /**
+     * @param Config $config
+     * @param StockConfigurationInterface $stockConfiguration
+     * @param StockIndexTableNameResolverInterface $stockIndexTableNameResolver
+     * @param ResourceConnection $resourceConnection
+     * @param StockByWebsiteIdResolverInterface $stockByWebsiteIdResolver
+     */
+    public function __construct(
+        Config $config,
+        StockConfigurationInterface $stockConfiguration,
+        StockIndexTableNameResolverInterface $stockIndexTableNameResolver,
+        ResourceConnection $resourceConnection,
+        StockByWebsiteIdResolverInterface $stockByWebsiteIdResolver
+    ) {
+        $this->config = $config;
+        $this->stockConfiguration = $stockConfiguration;
+        $this->stockIndexTableNameResolver = $stockIndexTableNameResolver;
+        $this->resourceConnection = $resourceConnection;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+    }
+
+    /**
+     * Filter out of stock options for configurable product.
+     *
+     * @param DataProvider $dataProvider
+     * @param array $indexData
+     * @param array $productData
+     * @param int $storeId
+     * @return array
+     * @SuppressWarnings(PHPMD.UnusedFormalParameter)
+     */
+    public function beforePrepareProductIndex(
+        DataProvider $dataProvider,
+        array $indexData,
+        array $productData,
+        int $storeId
+    ): array {
+        if ($this->config->isElasticsearchEnabled() && !$this->stockConfiguration->isShowOutOfStock($storeId)) {
+            $productIds = array_keys($indexData);
+            $stockStatuses = [];
+
+            foreach ($this->getProductIdsByWebsiteIds($productIds) as $websiteId => $productIds) {
+                $stock = $this->stockByWebsiteIdResolver->execute($websiteId);
+                $stockTable = $this->stockIndexTableNameResolver->execute((int)$stock->getStockId());
+
+                if ($this->resourceConnection->getConnection()->isTableExists($stockTable)) {
+                    $connection = $this->resourceConnection->getConnection();
+                    $select = $connection->select();
+                    $select->from(
+                        ['product' => $this->resourceConnection->getTableName('catalog_product_entity')],
+                        ['entity_id']
+                    );
+                    $select->joinInner(
+                        ['stock' => $stockTable],
+                        "product.sku = stock.sku",
+                        ['is_salable']
+                    );
+                    $select->where('product.entity_id IN (?)', $productIds);
+                    $productsSalable = $connection->fetchAssoc($select);
+                    $productsSalable = array_filter($productsSalable, function ($productSalable) {
+                        return StockStatusInterface::STATUS_IN_STOCK == $productSalable['is_salable'];
```

Comment:
always use === for comparison

Reply:
why not use filtering as a part of MySQL query?
why do you filter by `in stock` in application?